singleagent/contrastive_dyna.py

```diff
@@ -0,0 +1,435 @@
+        data.action[:-1],    # [T+1] --> [T]
+        self.extract_q(target_preds)[1:],  # [T+1] --> [T]
+        selector_actions[1:],  # [T+1] --> [T]
+        rewards[:-1],        # [T+1] --> [T]
+        discounts[:-1])      # [T+1] --> [T]
+
+    mask = data.discount[:-1]  # if 0, episode ending
+    batch_loss = 0.5 * jnp.square(batch_td_error).mean(axis=0)
+
+    batch_loss = vbb.maked_mean(batch_loss, mask)
+
+    metrics = {
+        '0.q_loss': batch_loss.mean(),
+        '0.q_td': jnp.abs(batch_td_error).mean(),
+        '1.reward': rewards.mean(),
+        'z.q_mean': self.extract_q(online_preds).mean(),
+        'z.q_var': self.extract_q(online_preds).var(),
+        }
+
+    return batch_td_error, batch_loss, metrics  # [T-1, B], [B]
+
+#!
+# class Predictions(NamedTuple):
+#     q_vals: jax.Array
+#     rnn_states: jax.Array
+@chex.dataclass(frozen=True)
+class Predictions:
+  state: jax.Array
+  q_values: jax.Array
+  rewards: jax.Array
+
+#! modified the following based on DummyRNN in networks.py in old code base
+class DummyRNN(nn.Module):
+    @nn.compact
+    def __call__(self, inputs: jax.Array, prev_state: Any
+                ) -> Tuple[jax.Array, Tuple[jax.Array, Any]]:
+        #? in old codebase we were using LSTMState, but I am not sure if we want that here
+        # so i use Any as a placeholder here
+        # Simply returns the inputs and previous state unchanged
+        return inputs, prev_state
+
+    def initial_state(self, batch_size: Optional[int] = None) -> Any:
+        # Generates an initial state for the RNN, which is a tuple of two zero arrays.
+        state_shape = (batch_size, 1) if batch_size is not None else (1,)
+        return jnp.zeros(state_shape)
+
+#! modified the SimpleTransition from old codebase (written in haiku)
+# need to use this for the transition function
+class SimpleTransition(nn.Module):
+    num_blocks: int
+
+    @nn.compact
+    def __call__(self, action_onehot: jnp.ndarray, prev_state: jnp.ndarray) -> jnp.ndarray:
+        channels = prev_state.shape[-1]
+        prev_state = nn.relu(prev_state)
+
+        encoded_action = nn.Dense(features=channels, use_bias=False)(action_onehot)
+        x_and_h = prev_state + encoded_action
+        out = nn.Sequential([nn.Dense(features=channels, use_bias=False) for _ in range(self.num_blocks)])(x_and_h)
+        return out
+
+#! copied this from old codebase utils.py cuz transition needs this
+# might consider adding this to the library but not sure which file
+def scale_gradient(g: jax.Array, scale: float) -> jax.Array:
+    """Scale the gradient.
+
+    Args:
+        g (_type_): Parameters that contain gradients.
+        scale (float): Scale.
+
+    Returns:
+        Array: Parameters with scaled gradients.
+    """
+    return g * scale + jax.lax.stop_gradient(g) * (1.0 - scale)
+
+#! modified the following from old codebase transition_fn
+# I actually don't understand why the original code returned tuple (out, out)
```

Comment:
We had `out, out` because the Transition function is in the "recurrent" category of functions, and these functions often have two outputs (1) the internal state (2) the output. For this special type of recurrent function, they happen to be the same thing. Using this formatting, makes it easy to use this module with utilities built for recurrent computation such as `jax.lax.scan`